gi-gtk-declarative-app-simple/src/GI/Gtk/Declarative/App/Simple.hs

```diff
@@ -59,14 +59,16 @@ data Transition state event =
 run
   :: Typeable event
   => App state event      -- ^ Application to run
-  -> IO ()
+  -> IO state
 run app = do
   void $ Gtk.init Nothing
-  void . async $ do
-    runLoop app
+  lastState <- async $ do
+    r <- runLoop app
     -- In case the run loop exits, quit the main GTK loop.
     Gtk.mainQuit
+    return r
   Gtk.main
+  wait lastState
```

Comment:
If `Gtk.main` exits before the `runLoop` (which shouldn't happen, but it can, theoretically), I'm thinking this last `wait lastState` will dead-lock. Maybe we should have an `MVar state`, initialized with `newEmptyMVar` before spawning the app loop, that will be written after `runLoop` exits, and instead of `wait lastState` we would do `tryTakeMVar` at the end, returning a `IO (Maybe state)`. What do you think?

Reply:
You are totally right, someone could call Gtk.mainQuit from anywhere and it should stop Gtk.main.

I think that a better approach could be checking if the async finished or not using some async function, after some digging `async` provides `poll :: Async a -> IO (Maybe (Either e a))` where Nothing means the thread hasn't finished, Just Right means it finished successfully and Just Left means an exception was thrown, I think I can use that and forward the Nothing and Just Right cases and throw an exception in the Just Left case, does that sound like a good idea? I don't think introducing more state is necessary to solve this particular issue.

Also that would give me the change to kill the thread if it outlived Gtk.main which sounds like a good idea, if the user feels like calling Gtk.mainQuit for some reason.

Reply:
The` async` solutions sounds very reasonable. :+1: